### PR TITLE
Many updates to dependencies and improved github workflows 

### PR DESCRIPTION
**Pull request by vesameskanen**
- Node 16
- Yarn v3
- Lots of more

This PR includes significant contributions from  Jannis.

Should be merged when no urgent UI releases are planned, because some fine tuning might become necessary.

### CODE REVIEW

**Comment by optionsome** [comment]
I wasn't able to build docker image locally at least:
<details>
➤ YN0000: │ sharp@npm:0.31.3 STDERR gyp ERR! build error 
➤ YN0000: │ sharp@npm:0.31.3 STDERR gyp ERR! stack Error: `make` failed with exit code: 2
➤ YN0000: │ sharp@npm:0.31.3 STDERR gyp ERR! stack     at ChildProcess.onExit (/opt/digitransit-ui/node_modules/node-gyp/lib/build.js:203:23)
➤ YN0000: │ sharp@npm:0.31.3 STDERR gyp ERR! stack     at ChildProcess.emit (node:events:513:28)
➤ YN0000: │ sharp@npm:0.31.3 STDERR gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:293:12)
➤ YN0000: │ sharp@npm:0.31.3 STDERR gyp ERR! System Linux 4.19.0-6-amd64
➤ YN0000: │ sharp@npm:0.31.3 STDERR gyp ERR! command "/usr/local/bin/node" "/opt/digitransit-ui/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
➤ YN0000: │ sharp@npm:0.31.3 STDERR gyp ERR! cwd /opt/digitransit-ui/node_modules/sharp
➤ YN0000: │ sharp@npm:0.31.3 STDERR gyp ERR! node -v v16.20.1
➤ YN0000: │ sharp@npm:0.31.3 STDERR gyp ERR! node-gyp -v v9.4.0
➤ YN0000: │ sharp@npm:0.31.3 STDERR gyp ERR! not ok 
➤ YN0009: │ sharp@npm:0.31.3 couldn't be built successfully (exit code 1, logs can be found here: /tmp/xfs-58f43b73/build.log)
➤ YN0007: │ digitransit-ui@workspace:. must be built because it never has been before or the last one failed
</details>

**Comment by derhuerst** [comment]
> I wasn't able to build docker image locally at least:
> 
> ```
> ➤ YN0009: │ sharp@npm:0.31.3 couldn't be built successfully (exit code 1, logs can be found here: /tmp/xfs-58f43b73/build.log)
> ```

I cannot reproduce this on macOS 12.6.8 on an Apple Silicon M2, with Node v16.20.1 & the local Yarn v3.4.1.

@optionsome Can you provide the referenced log (or a new equivalent one)?
Which platform are you building on (OS, CPU architecture)?

**Comment by vesameskanen** [comment]
'docker build .' does not work in my mac, but I noticed that current v3 branch has the same problem. No clear error message, just exit code 129. Works when using github actions though.

workflow did not have  library publishing job at all, and I tried adding it following the pattern of other  jobs. Seems to work.

File: .github/workflows/hotfix-pipeline.yml [hunk 1]
```diff
@@ -14,9 +14,35 @@ jobs:
...
-        run: ./.github/workflows/scripts/build_and_push_dev.sh
-        env:
-          DOCKER_USER: ${{ secrets.DOCKER_USER }}
-          DOCKER_AUTH: ${{ secrets.DOCKER_AUTH }}
-          DOCKER_BASE_TAG: hotfix
+
+      - name: configure Docker Buildx
+        uses: docker/setup-buildx-action@v1
+      - name: log in to DockerHub
+        uses: docker/login-action@v1
```
Comment (optionsome) [inline .github/workflows/hotfix-pipeline.yml]:
```suggestion
        uses: docker/login-action@v2
```

File: .github/workflows/hotfix-pipeline.yml [hunk 2]
```diff
@@ -14,9 +14,35 @@ jobs:
...
           timezone: "Europe/Helsinki"
-      - name: Build docker image from next and push it
-        run: ./.github/workflows/scripts/build_and_push_dev.sh
-        env:
-          DOCKER_USER: ${{ secrets.DOCKER_USER }}
-          DOCKER_AUTH: ${{ secrets.DOCKER_AUTH }}
-          DOCKER_BASE_TAG: hotfix
+
+      - name: configure Docker Buildx
+        uses: docker/setup-buildx-action@v1
```
Comment (optionsome) [inline .github/workflows/hotfix-pipeline.yml]:
```suggestion
        uses: docker/setup-buildx-action@v2
```

File: .github/workflows/hotfix-pipeline.yml [hunk 3]
```diff
@@ -14,9 +14,35 @@ jobs:
...
+          path: '.yarn/cache'
+          key: yarn-cache-${{ github.ref_name }}
+          restore-keys: yarn-cache-v2
+
+      - name: determine Docker image tags
+        id: docker-tags
+        run: |
+          echo "::set-output name=permanent-tag::hotfix-$(date +"%Y-%m-%dT%H.%M.%S")-${GITHUB_SHA:0:7}"
+      - name: build and push Docker image
+        uses: docker/build-push-action@v2
```
Comment (optionsome) [inline .github/workflows/hotfix-pipeline.yml:38]:
```suggestion
        uses: docker/build-push-action@v4
```

File: Dockerfile [hunk 4]
```diff
@@ -29,17 +100,13 @@ ENV \
...
-ADD . ${WORK}
-
-RUN \
-  yarn && \
-  yarn setup && \
-  yarn build && \
-  rm -rf static docs test /tmp/* .cache && \
-  yarn cache clean --all
+RUN apk add --no-cache curl
+HEALTHCHECK \
```
Comment (optionsome) [inline Dockerfile:108]: I wonder what this healthcheck does in practice. If we have configured different health checks in kubernetes, what happens?

Comment (derhuerst) [reply Dockerfile:108]: The health check is picked up by the regular Docker CLI (`docker run`), Docker Swarm, and various other container-based tools.

[Kubernetes ignores it if there is a readiness/liveness check defined](https://stackoverflow.com/a/41476481).

File: Dockerfile [hunk 5]
```diff
@@ -104,9 +104,4 @@ ENV \
   ASSET_URL='' \
   STATIC_MESSAGE_URL=''
 
-RUN apk add --no-cache curl
-HEALTHCHECK \
```
Comment (derhuerst) [inline Dockerfile]: I strongly recommend to keep the `HEALTHCHECK`. It helps in a lot of use cases, and I'm not aware of any disadvantages.

For example, one wouldn't notice with a Docker-based dev setup that `digitransit-ui`'s web server actually (including the bundling it does) isn't working as intended; With the health check, this is reported to the person running `docker run`.

**Pull request merged**